tvb_contrib/tvb/contrib/cosimulation/cosim_monitors.py

```diff
@@ -0,0 +1,185 @@
+# -*- coding: utf-8 -*-
+#
+#
+#  TheVirtualBrain-Contributors Package. This package holds simulator extensions.
+#  See also http://www.thevirtualbrain.org
+#
+# (c) 2012-2020, Baycrest Centre for Geriatric Care ("Baycrest") and others
+#
+# This program is free software: you can redistribute it and/or modify it under the
+# terms of the GNU General Public License as published by the Free Software Foundation,
+# either version 3 of the License, or (at your option) any later version.
+# This program is distributed in the hope that it will be useful, but WITHOUT ANY
+# WARRANTY; without even the implied warranty of MERCHANTABILITY or FITNESS FOR A
+# PARTICULAR PURPOSE.  See the GNU General Public License for more details.
+# You should have received a copy of the GNU General Public License along with this
+# program.  If not, see <http://www.gnu.org/licenses/>.
+#
+#
+#   CITATION:
+# When using The Virtual Brain for scientific publications, please cite it as follows:
+#
+#   Paula Sanz Leon, Stuart A. Knock, M. Marmaduke Woodman, Lia Domide,
+#   Jochen Mersmann, Anthony R. McIntosh, Viktor Jirsa (2013)
+#       The Virtual Brain: a simulator of primate brain network dynamics.
+#   Frontiers in Neuroinformatics (7:10. doi: 10.3389/fninf.2013.00010)
+
+"""
+.. moduleauthor:: Lionel Kusch <lkusch@thevirtualbrain.org>
+.. moduleauthor:: Dionysios Perdikis <dionperd@gmail.com>
+"""
+
+import abc
+
+import numpy
+
+from tvb.basic.neotraits.api import HasTraits, Attr, NArray
+from tvb.simulator.coupling import Coupling, Linear
+from tvb.simulator.monitors import Raw, RawVoi, AfferentCoupling
+
+
+class CosimMonitor(HasTraits):
+    """
+    Abstract base class for cosimulation monitors implementations.
+    """
+
+    def get_sample(self, start_step, n_steps, history,cosim):
+        times = []
+        values = []
+        for step in range(start_step, start_step + n_steps):
+            if cosim:
+                state = history.query(step)
```

Comment:
For repeated calls to query your cosim history, why not just implement a new method on that history instance to retrieve data for the `start_step` to `start_step+n_steps` interval at once?

Reply:
I don't really see the point here since anyway we need to loop for each step in case history is a normal TVB history and in all cases for the sampling with the TVB monitor.

Reply:
There are a need to include this new function to the scientific library.
We doesn't include it for the moment.
We can do it if you want.
if yes, what name do you propose for this new function?

Reply:
no it's ok, was minor comment